File: scripts/sweep_v4.mjs
#!/usr/bin/env node
/**
 * Sweep Stage-2 configs on canonical per-line corpus + mixed eval.
 *
 * Trains on corpus/bpe_train.txt split into LINES (boundary-preserving),
 * then reports BOTH benchmark-8 metrics AND held-out generalization metrics.
 * Selection: maximize min(heldout win) first, then benchmark total win.
 */
import { readFileSync } from 'fs';
import { encode } from '../src/encoder.js';
import { decode } from '../src/decoder.js';
import { trainTokenizer, tokenize, detokenize } from '../src/tokenizer/index.js';
import { encode as gptEncode } from 'gpt-tokenizer';

const BENCH = [
  ['Common English', 'the quick brown fox jumps over the lazy dog this is a test of the emergency broadcast system how now brown cow the rain in spain stays mainly on the plain'],
  ['Code', 'const app = express(); app.get("/api/tasks", async (req, res) => { const tasks = await db.query("SELECT * FROM tasks"); res.json(tasks); });'],
  ['SQL', 'SELECT * FROM users WHERE id=42 AND name LIKE \'%test%\' ORDER BY created_at DESC; INSERT INTO table_name (a,b,c) VALUES (1,\'x\',true); UPDATE users SET name=\'abc\', score=99 WHERE id=7; DELETE FROM users WHERE id>1000; CREATE TABLE test(id INT PRIMARY KEY,name VARCHAR(255));'],
  ['API', '{"status": "success", "data": {"users": [{"id": 1, "name": "John", "email": "john@example.com"}, {"id": 2, "name": "Jane", "email": "jane@example.com"}], "total": 2, "page": 1, "per_page": 10}}'],
  ['Shell', '$ echo "Hello, World!"; printf \'%s\\n\' "$HOME"; ls -la /tmp | grep ".log" && cat file.txt; sudo -n true || echo "no sudo"; python3 -c \'print("test")\'; node -e "console.log(42)"; git status --short; git add . && git commit -m "test"; npm run build && npm start'],
  ['Markdown', '# README.md ## Test Project ### Features - fast - simple - random ### Code `npm install && npm run dev` **bold** *italic* [link](https://example.com) > quote --- ### End'],
  ['Paths', 'C:\\Users\\Test\\file.txt D:\\Games\\MC\\server.exe /usr/bin/bash ~/.config/hypr/hyprland.conf ../../src/main.js ./build/output.log https://example.com/?a=1&b=2 ftp://x@y.z:21/path git@host:user/repo.git user@example.com test+tag@example.org'],
  ['Prompt', 'aicl is Goated BTW, and this can reduce tokens very vary fast'],
];

// NEVER in training corpus — pure generalization probe
const HELDOUT = [
  'yesterday evening the deployment pipeline failed because the database migration timed out twice',
  'the quarterly revenue report shows significant growth across all regional markets this year',
  'function calculateTotal(items, taxRate) { return items.reduce((s, x) => s + x.price, 0) * (1 + taxRate); }',
  'every morning the engineer reviews open pull requests before merging anything into main',
  'INSERT INTO inventory (sku, quantity, warehouse) VALUES ($1, $2, $3) RETURNING id;',
  'curl -fsSL https://releases.example.org/v2.tar.gz | tar -xz && ./install --prefix ~/.local',
];

const lines = readFileSync('corpus/bpe_train.txt', 'utf-8').split('\n').filter(Boolean);
console.log(`corpus lines: ${lines.length}, chars: ${lines.join('').length}`);

function evalSet(vocab, set) {
  let raw = 0, aicl = 0, tok = 0, gpt = 0, minWin = Infinity, lossless = true;
  const rows = [];
  for (const [name, text] of set) {
    const enc = encode(text);
    const a = [...enc.output].length;
    const ids = tokenize(enc.output, vocab);
    const rt = decode(detokenize(ids, vocab)).output === text;
    if (!rt) lossless = false;
    const g = gptEncode(text, { model: 'gpt-4o' }).length;
    const w = g / ids.length;
    if (w < minWin) minWin = w;
    raw += [...text].length; aicl += a; tok += ids.length; gpt += g;
    rows.push({ name, cpt: (a / ids.length).toFixed(2), win: w.toFixed(2), rt });
  }
  return { raw, aicl, tok, gpt, cpt: aicl / tok, win: gpt / tok, minWin, lossless, rows };
}

const CONFIGS = [
  { numMerges: 1024, maxTokenLength: 8, minFrequency: 2 },
  { numMerges: 2048, maxTokenLength: 8, minFrequency: 3 },
  { numMerges: 2048, maxTokenLength: 10, minFrequency: 3 },
  { numMerges: 4096, maxTokenLength: 10, minFrequency: 5 },
];

const results = [];
for (const cfg of CONFIGS) {
  const t0 = Date.now();
  const vocab = trainTokenizer(lines, { ...cfg, mergeBase: 100000 });
  const ms = Date.now() - t0;
  const b = evalSet(vocab, BENCH);
  const h = evalSet(vocab, HELDOUT.map((t, i) => ['held' + i, t]));
  results.push({ cfg, actual: vocab.numMerges, ms, b, h });
  console.log(`merges=${cfg.numMerges} maxLen=${cfg.maxTokenLength} minFreq=${cfg.minFrequency} actual=${vocab.numMerges} ${ms}ms`);
  console.log(`  bench: CPT=${b.cpt.toFixed(2)} win=${b.win.toFixed(2)}x minWin=${b.minWin.toFixed(2)} lossless=${b.lossless}`);
  console.log(`  held : CPT=${h.cpt.toFixed(2)} win=${h.win.toFixed(2)}x minWin=${h.minWin.toFixed(2)} lossless=${h.lossless}`);
}

console.log('\n' + '='.repeat(90));
console.log('SWEEP (bench CPT/win + held-out CPT/win)');
console.log('='.repeat(90));
for (const r of results) {
  console.log(`req=${r.cfg.numMerges} act=${r.actual} len=${r.cfg.maxTokenLength} | bench CPT=${r.b.cpt.toFixed(2)} win=${r.b.win.toFixed(2)}x min=${r.b.minWin.toFixed(2)} | held CPT=${r.h.cpt.toFixed(2)} win=${r.h.win.toFixed(2)}x min=${r.h.minWin.toFixed(2)} | ${r.ms}ms`);
  for (const row of r.b.rows) console.log(`    bench ${row.name}: CPT=${row.cpt} win=${row.win}x rt=${row.rt ? 'ok' : 'FAIL'}`);
  for (const row of r.h.rows) console.log(`    held  ${row.name}: CPT=${row.cpt} win=${row.win}x rt=${row.rt ? 'ok' : 'FAIL'}`);
}

File: scripts/sweep_v4.py
#!/usr/bin/env python3
"""Sweep canonical Stage-2 configurations on the v4 line corpus."""

from __future__ import annotations

import argparse
import sys
import time
from pathlib import Path
from typing import Any, Sequence

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

try:
    from .eval_vocab import _tokenizer_api, evaluate, vocab_summary  # noqa: E402
    from .heldout_probes import BENCH  # noqa: E402
    from .train_fast import read_lines  # noqa: E402
except ImportError:
    from eval_vocab import _tokenizer_api, evaluate, vocab_summary  # noqa: E402
    from heldout_probes import BENCH  # noqa: E402
    from train_fast import read_lines  # noqa: E402

HELDOUT = [
    ("held0", "yesterday evening the deployment pipeline failed because the database migration timed out twice"),
    ("held1", "the quarterly revenue report shows significant growth across all regional markets this year"),
    ("held2", "function calculateTotal(items, taxRate) { return items.reduce((s, x) => s + x.price, 0) * (1 + taxRate); }"),
    ("held3", "every morning the engineer reviews open pull requests before merging anything into main"),
    ("held4", "INSERT INTO inventory (sku, quantity, warehouse) VALUES ($1, $2, $3) RETURNING id;"),
    ("held5", "curl -fsSL https://releases.example.org/v2.tar.gz | tar -xz && ./install --prefix ~/.local"),
]
DEFAULT_CONFIGS = (
    {"num_merges": 1024, "max_token_length": 8, "min_frequency": 2},
    {"num_merges": 2048, "max_token_length": 8, "min_frequency": 3},
    {"num_merges": 2048, "max_token_length": 10, "min_frequency": 3},
    {"num_merges": 4096, "max_token_length": 10, "min_frequency": 5},
)


def _train_api() -> Any:
    try:
        from aicl.tokenizer import train_tokenizer
    except ImportError:
        from aicl import train_tokenizer  # type: ignore[attr-defined]
    return train_tokenizer


def self_check() -> None:
    assert len(HELDOUT) == 6
    assert DEFAULT_CONFIGS[0]["num_merges"] == 1024
    print("sweep_v4 self-check: ok")


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--corpus", type=Path, default=ROOT / "corpus" / "bpe_train.txt")
    parser.add_argument("--num-merges", type=int)
    parser.add_argument("--max-token-length", type=int)
    parser.add_argument("--min-frequency", type=int)
    parser.add_argument("--self-check", action="store_true")
    args = parser.parse_args(argv)
    if args.self_check:
        self_check()
        return 0
    lines = read_lines(args.corpus)
    print(f"corpus lines: {len(lines)}, chars: {sum(map(len, lines))}")
    train_tokenizer = _train_api()
    _load, tokenize, detokenize = _tokenizer_api()
    from aicl import decode

    results: list[dict[str, Any]] = []
    for original in DEFAULT_CONFIGS:
        config = dict(original)
        if args.num_merges is not None:
            config["num_merges"] = args.num_merges
        if args.max_token_length is not None:
            config["max_token_length"] = args.max_token_length
        if args.min_frequency is not None:
            config["min_frequency"] = args.min_frequency
        started = time.perf_counter()
        vocab = train_tokenizer(
            lines,
            {
                "num_merges": config["num_merges"],
                "merge_base": 100_000,
                "max_token_length": config["max_token_length"],
                "min_frequency": config["min_frequency"],
            },
        )
        elapsed = (time.perf_counter() - started) * 1000
        bench = evaluate(BENCH, vocab, tokenize=tokenize, detokenize=detokenize, decode=decode)
        held = evaluate(HELDOUT, vocab, tokenize=tokenize, detokenize=detokenize, decode=decode)
        result = {"config": config, "actual": vocab_summary(vocab)[0], "ms": elapsed, "bench": bench, "held": held}
        results.append(result)
        print(
            f"merges={config['num_merges']} maxLen={config['max_token_length']} "
            f"minFreq={config['min_frequency']} actual={vocab_summary(vocab)[0]} {elapsed:.0f}ms "
            f"bench={bench['tokens']} held={held['tokens']} lossless={bench['lossless'] and held['lossless']}"
        )
    print("\nSWEEP (Stage-2 CPT/win without optional GPT baselines)")
    for result in results:
        config = result["config"]
        print(
            f"req={config['num_merges']} act={result['actual']} len={config['max_token_length']} | "
            f"bench CPT={result['bench']['cpt']:.2f} | held CPT={result['held']['cpt']:.2f} | {result['ms']:.0f}ms"
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
